lib/spack/docs/tutorial_modules.rst

```diff
@@ -1,52 +1,278 @@
+    module files. It's a drop-in replacement of Environment Modules and works with
+    both of the module file formats generated by Spack.
+    Despite being fully compatible with Environment Modules there are many features that
+    are unique to Lmod. These features are either
+    `targeted to safety <http://lmod.readthedocs.io/en/latest/010_user.html#safety-features>`_
+    or meant to
+    `extend the module system functionality <http://lmod.readthedocs.io/en/latest/010_user.html#module-hierarchy>`_.
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How do we generate module files?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Before  we dive into the hands-on it's worth spending a couple of words to explain how
+module files are generated by Spack. The following diagram provides a high-level view
+of the process:
+
+
+.. image:: module_file_generation.*
+
+The red dashed line above represents Spack's boundaries, the blue one Spack's vendored
+dependencies. In short, module files are generated combining:
+
+  * the configuration details in ``config.yaml`` and ``modules.yaml``
+  * the information contained in Spack packages (and processed by the module subpackage)
+  * a set of `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_ templates
+
+with the vendored template engine to "stamp out" each particular module file. As Spack serves
+very diverse needs this process has many points of customization, and we'll explore most of
+them in the next sections.
 
 .. _module_file_tutorial_prerequisites:
 
--------------
-Prerequisites
--------------
+-----------------------
+Set up for the tutorial
+-----------------------
+
+In order to showcase the capabilities of Spack's module file generation, we need
+a representative set of software to work with. This set must include different
+flavors of the same packages installed alongside each other and some
+:ref:`external packages <sec-external-packages>`.
 
-Before proceeding further ensure:
+The purpose of this setup is not to make our life harder but to demonstrate
+how Spack can help with similar situations, as they will happen on real HPC clusters.
+For instance, it's often preferable for Spack to use vendor-provided MPI
+implementations than to build one itself.
 
-- you have LMod or Environment Modules available
-- have :ref:`shell support <shell-support>` activated in Spack
+The best way to follow along is to use a Docker image, which comes
+with Spack and all the software used in the following parts already
+pre-installed. If you want to proceed this way, read :ref:`module_file_tutorial_use_docker`.
 
-If you need to install Lmod or Environment module you can refer
-to the documentation :ref:`here <InstallEnvironmentModules>`.
+If you don't have Docker installed or for any other reason you
+prefer to work locally, follow instead :ref:`module_file_tutorial_work_locally`
+to know how to clone Spack and install the software.
+Be aware that the set-up will take longer and that the details of the snippets
```

Comment:
I think "longer" is an understatement. The last time I tried to build the full set of software in the tutorial with Spack-alone, it took 3 hours. I see that you're offloading some of this on the system package manager. Many users might not even have access to that. I'm not sure if I can think of a better solution than telling them to use Docker.

Reply:
> Many users might not even have access to that. 

The manual setup must be taken as a guideline, not as a set of cut and paste commands. There are  far too many possible options to cover if we go that way. For instance, users on a cluster  won't have access to a system package manager, but they probably have already an external MPI installed and can use that in `packages.yaml`.

Reply:
> The last time I tried to build the full set of software in the tutorial with Spack-alone, it took 3 hours

Confirmed: almost 3 hours to rebuild the image on my laptop.